### PR TITLE
Update Select component

### PR DESCRIPTION
**Pull request by timoheddes**
### Select Component updates

[Asana Link](https://app.asana.com/0/1204958723576385/1205164815746147/f)

**What changed?**
- Updated Select styling to match design and support dark and light theme
- Tweaked the size of the native HTML select element that we render to ensure you can click anywhere on the actual Select component to see its options (previously required to click in the center of our visual component)
- Added disabled styling (missing exact design, [also missing in e.g. Input component](https://app.asana.com/0/1205288273408141/1205332525545630/f)?)
- Add variant prop to allow switching between default form select and one with a solid border and transparent background
- Set Tooltip z-index as a temporary fix to ensure it's above form elements (created [task](https://app.asana.com/0/1205288273408141/1205307366740459/f) to fix this for all components)
- Removed style from option elements as these can't be styled (?? not sure what the idea was there, am I missing something?)
- Added typing to the Input component type to match valid HTML Input Types rather than any string and updated Storybook implementation to allow switching between input types

**Implementation preview in Storybook, showing Input and Select:**

https://github.com/kadena-community/kadena.js/assets/79908211/75a157ed-c58f-4001-8095-0162d07a50d5

### CODE REVIEW

File: packages/libs/react-ui/src/components/Select/Select.tsx
Comment (eileenmguo) [inline packages/libs/react-ui/src/components/Select/Select.tsx:55]: why are you wrapping this in a label?

File: packages/libs/react-ui/src/components/Select/Select.tsx
Comment (eileenmguo) [inline packages/libs/react-ui/src/components/Select/Select.tsx:28]: You can remove "variants" and type it with just the variant object 
```suggestion
    variant?: keyof typeof selectVariants;
```

**Review by eileenmguo — COMMENTED** [review]
Looking at storybook, the form version of the select is very close but doesn't look aligned with the input component. The idea is that these would be used together, so any visual inconsistencies will be very apparent. I would suggest for testing purposes to just put them on top of each other in storybook to compare styling during development

File: packages/libs/react-ui/src/components/Select/Select.tsx
Comment (timoheddes) [reply packages/libs/react-ui/src/components/Select/Select.tsx:55]: It's just so that clicking the (optional) icon will also highlight the select field.

File: packages/libs/react-ui/src/components/Select/Select.css.ts
Comment (timoheddes) [reply packages/libs/react-ui/src/components/Select/Select.css.ts:51]: I guess it's between use case vs what to expect visually. I think it's the default one that would be used in a form though, right? The one I named solid shows up in the design as part of the header, so default would become form and the solid one would become default? But that doesn't make too much sense either. I initially used default and form but found it confusing. What do you think? Perhaps I'm wrong thinking that the current default one would be the form one..

File: packages/libs/react-ui/src/components/Select/Select.css.ts
Comment (timoheddes) [reply packages/libs/react-ui/src/components/Select/Select.css.ts:36]: No :)

File: packages/libs/react-ui/src/components/Select/Select.tsx
Comment (timoheddes) [reply packages/libs/react-ui/src/components/Select/Select.tsx]: I updated it now to make the original wrapping div the label, so we don't have another depth added to it.

**Comment by timoheddes** [comment]
> Looking at storybook, the form version of the select is very close but doesn't look aligned with the input component. The idea is that these would be used together, so any visual inconsistencies will be very apparent. I would suggest for testing purposes to just put them on top of each other in storybook to compare styling during development

I changed it to match input but I noticed that with the input component, its height jumps when an icon is added. I assume this is not the intention? Do you know? Otherwise I'll check with Isa. I can update the input component to remain the same height with or without the icon as well. Seems strange to me to have variation in the input heights in a form depending on whether or not an icon is added.

**Comment by MRVDH** [comment]
We have select boxes in the Tools project header: https://www.figma.com/file/A11XXPwJSHNSqrRBEFfRc8/K%3AApps?type=design&node-id=725-11387&mode=design&t=wcBXEna6ES40ftxc-0
image

But since the regular form select is different, should this select in the header be a custom component? Is this something we should check with @isaozler?

cc: @KristinaSpasevska

**Comment by timoheddes** [comment]
@MRVDH part of this PR is adding a variant, which would be the one you're linking to there :)

File: packages/libs/react-ui/src/components/Select/Select.css.ts
Comment (eileenmguo) [reply packages/libs/react-ui/src/components/Select/Select.css.ts:50]: hmmm I see where you're coming from. So we can set the default variant prop to anything regardless of what we name it. I just think the form variant should be named form for clarity and then we could name the other one something descriptive as well. I do think you're right that it will most likely be used in a form so that can be the default variant in the end

File: packages/libs/react-ui/src/components/Select/Select.tsx
Comment (eileenmguo) [reply packages/libs/react-ui/src/components/Select/Select.tsx]: oh interesting, I didn't realize it worked like that. Do you know how that effects accessibility? Also we use our `InputWrapper` to add an additional label?

Comment (timoheddes) [reply packages/libs/react-ui/src/components/Select/Select.tsx]: Ahh ok, I didn't realise that. Then it's a label wrapped in a label now and we don't want that 😁 I'll change it back in that case!

My change came from the comment from Maarten who said the "Click listener only listens on text." which he explained to mean that if he clicks outside of the select box, like on the icon, it won't trigger the select box. But if we wrap the whole thing in a label through the inputWrapper, that shouldn't be an issue once it's implemented in the form as the label should trigger the corresponding form item.

Comment (eileenmguo) [reply packages/libs/react-ui/src/components/Select/Select.tsx]: No, there would not be a label wrapped around it. It would be adjacent to it and connected via htmlFor. So like 
```
<label>
<label/>
<label>
  <select/>
</label>
```

Comment (eileenmguo) [reply packages/libs/react-ui/src/components/Select/Select.tsx]: Can you achieve this by just having the select span the entire container, adding extra padding on the left, and positioning the icon absolutely with no pointerevents?

Comment (timoheddes) [reply packages/libs/react-ui/src/components/Select/Select.tsx]: If the for tag on the label matches the id of the input, it works the same as wrapping it around

File: packages/libs/react-ui/src/components/Select/Select.css.ts
Comment (timoheddes) [reply packages/libs/react-ui/src/components/Select/Select.css.ts]: So now the default = form and we have flat as alternative. I'll resolve this conversation but we'll get approval from all parties involved before merging :)

File: packages/libs/react-ui/src/components/Select/Select.css.ts
Comment (timoheddes) [inline packages/libs/react-ui/src/components/Select/Select.css.ts:49]: Had to add important to overwrite the dark theme default background styling. Please let me know if there's another better way.

Comment (eileenmguo) [reply packages/libs/react-ui/src/components/Select/Select.css.ts:49]: You should be able to use sprinkles for this. I think it might work without the !important that way

Comment (timoheddes) [reply packages/libs/react-ui/src/components/Select/Select.css.ts:46]: I tried just now but it continues to overwrite it without adding important.

Comment (eileenmguo) [reply packages/libs/react-ui/src/components/Select/Select.css.ts:46]: Maybe it's because you are applying the backgroundColor in the containerClass which has more specificity. I would also suggest moving any variant specific styling to the variants styling section only only keeping shared styling in the containerClass. I feel like this may fix the issue

File: packages/libs/react-ui/src/components/Select/Select.stories.tsx
Comment (eileenmguo) [inline packages/libs/react-ui/src/components/Select/Select.stories.tsx:54]:
```suggestion
    variant: keyof typeof SelectVariants;
```

File: packages/libs/react-ui/src/components/Select/Select.stories.tsx
Comment (eileenmguo) [inline packages/libs/react-ui/src/components/Select/Select.stories.tsx:63]:
```suggestion
  render: ({ icon, disabled, variant = "form" }) => {
```

File: packages/libs/react-ui/src/components/Select/Select.tsx
Comment (eileenmguo) [inline packages/libs/react-ui/src/components/Select/Select.tsx:16]: I think this can be removed

File: packages/libs/react-ui/src/components/Select/Select.tsx
Comment (eileenmguo) [inline packages/libs/react-ui/src/components/Select/Select.tsx:42]:
```suggestion
    variant = "form",
```

Comment (timoheddes) [reply packages/libs/react-ui/src/components/Select/Select.tsx:42]: Wait what why do you suggest this? It would mean that you have to change it multiple places if we decide to change the variant name. Having this default to the first option within the variants makes sense I think?

File: packages/libs/react-ui/src/components/Select/Select.tsx
Comment (timoheddes) [reply packages/libs/react-ui/src/components/Select/Select.tsx:16]: I did that so there's one source of truth and I can use it with the `variants[0]`, so we only have to change it in one location if we ever do.

File: packages/libs/react-ui/src/components/Select/Select.css.ts
Comment (timoheddes) [reply packages/libs/react-ui/src/components/Select/Select.css.ts]: Ah yes, ok so I changed it to not have the background set on the containerClass and that works.

File: packages/libs/react-ui/src/components/Input/Input.stories.tsx
Comment (eileenmguo) [inline packages/libs/react-ui/src/components/Input/Input.stories.tsx:156]: Did you mean to leave this?

File: packages/libs/react-ui/src/components/Select/Select.css.ts
Comment (eileenmguo) [inline packages/libs/react-ui/src/components/Select/Select.css.ts:30]: What was the reason for changing the disabled state?

File: packages/libs/react-ui/src/components/Select/Select.tsx
Comment (eileenmguo) [inline packages/libs/react-ui/src/components/Select/Select.tsx:48]: nit: I would order the classNames in order of specificity. So since containerClass is the most generic, it should go first so that it can be overridden (I think) by the following classnames. This might be why you needed to remove the backgroundColor from containerClass

Just a note for the future - might help with styling!

File: packages/libs/react-ui/src/components/Select/Select.tsx
Comment (eileenmguo) [inline packages/libs/react-ui/src/components/Select/Select.tsx:14]: I'm not opposed to having a selectVariant type, but as discussed I think it should be based on the actual variant object rather than something that you define yourself. I believe this would be `keyof typeof selectVariants`

File: packages/libs/react-ui/src/components/Select/Select.tsx
Comment (eileenmguo) [inline packages/libs/react-ui/src/components/Select/Select.tsx:28]: If you write it this way, you can remove the SelectVariants type
```suggestion
  variant?: keyof typeof selectVariants;
```

File: packages/libs/react-ui/src/components/Select/Select.css.ts
Comment (timoheddes) [reply packages/libs/react-ui/src/components/Select/Select.css.ts:30]: I made a [task](https://app.asana.com/0/1205288273408141/1205332525545630/f) on the board but there appears to be no disabled state implemented, even though it appears to exist in the design. If you look in Storybook, you see that the input component currently has no visual indication that it's disabled.

Comment (eileenmguo) [reply packages/libs/react-ui/src/components/Select/Select.css.ts:30]: Right. So right now the `Input` doesn't have a disabled style (which is a bug), however the `InputWrapper` and subsequently the `TextField` do (because of the `InputWrapper`) which I believe is similar to the original implementation for this disabled state. I would look there to decide how to style this disable state

Note: I saw that the input component styles look off now (leading text is on top of the actual input?)

Comment (timoheddes) [reply packages/libs/react-ui/src/components/Select/Select.css.ts:30]: What is this? I figured that's disabled. Though the design isn't too clear about it. In any case, wouldn't it make a lot more sense to have the disabled styling on the actual element rather than managed by a wrapper element? Or is that not what you're saying about the current situation?

image

Ah yeah I see the leading text is off now, I'll fix that.

Comment (eileenmguo) [reply packages/libs/react-ui/src/components/Select/Select.css.ts:29]: Hmm to be honest, I'm not sure what that state is. I'm saying when the wrapper component is disabled, it applies the disabled style to everything, but also the input needs to be updated to have a disabled state. @isaozler is back so let's have him take a look and we can decide if we want to break this into a different ticket?

Comment (eileenmguo) [reply packages/libs/react-ui/src/components/Select/Select.css.ts:29]: Actually I think it's fine to already break this into a separate ticket since the input component also needs updates

Comment (timoheddes) [reply packages/libs/react-ui/src/components/Select/Select.css.ts:29]: Yeah I think that would be best!

**Review by EnoF — APPROVED** [review]

**Review by eileenmguo — COMMENTED** [review]
border radius and colors are off on the `transparent` variant. I would suggest adding it to the nav header to see if it looks right

image

File: packages/libs/react-ui/src/components/Select/Select.tsx
Comment (timoheddes) [reply packages/libs/react-ui/src/components/Select/Select.tsx]: Yeah 👍

**Review by eileenmguo — APPROVED** [review]

**Pull request merged**
